lib/model/indexhandler.go

```diff
@@ -328,12 +368,66 @@ func (s *indexHandler) receive(fs []protocol.FileInfo, update bool, op string) e
 	if !update {
 		fset.Drop(deviceID)
 	}
+
+	l.Debugf("Received %d files for %s from %s, prevSeq=%d, lastSeq=%d", len(fs), s.folder, deviceID.Short(), prevSequence, lastSequence)
+
+	// Verify that the previous sequence number matches what we expected
+	if exp := fset.Sequence(deviceID); prevSequence > 0 && prevSequence != exp {
+		s.logSequenceAnomaly("index update with unexpected sequence", map[string]any{
+			"prevSeq":      prevSequence,
+			"lastSeq":      lastSequence,
+			"batch":        len(fs),
+			"expectedPrev": exp,
+		})
+	}
+
 	for i := range fs {
+		// Verify index in relation to the claimed sequence boundaries
+		if fs[i].Sequence < prevSequence {
+			s.logSequenceAnomaly("file with sequence before prevSequence", map[string]any{
+				"prevSeq": prevSequence,
+				"lastSeq": lastSequence,
+				"batch":   len(fs),
+				"seenSeq": fs[i].Sequence,
+				"atIndex": i,
+			})
+		}
+		if lastSequence > 0 && fs[i].Sequence > lastSequence {
+			s.logSequenceAnomaly("file with sequence after lastSequence", map[string]any{
+				"prevSeq": prevSequence,
+				"lastSeq": lastSequence,
+				"batch":   len(fs),
+				"seenSeq": fs[i].Sequence,
+				"atIndex": i,
+			})
+		}
```

Comment:
Isn't this always going to fire? Next is always bigger than last, no?

Reply:
Confusion of English terminology, sorry. "last" as "the supposed last sequence number of the batch we're currently processing".